### DISCUSSION TITLE
Can't find GDM and Polkit using Flatpak on NixOS

### DISCUSSION THREAD

**Discussion by RossComputerGuy** [realmazharhussain/gdm-settings#179] [General]
### What happened?

I get a missing dependencies error with the Flatpak on NixOS.

### Expected behavior

It should find the dependencies

### To Reproduce

Install nixOS with GDM and Polkit and install GDM Settings through Flatpak

### Application Version

stable

### Source

Flatpak

### Environment

- **OS**: NixOS 22.05
- **GNOME**: *n/a*

**Comment by realmazharhussain** [comment]
What is the output of the following command on your system?

```bash
pkexec --version; echo Return Status: $?
```

**Comment by RossComputerGuy** [comment]
```
pkexec version 0.120
Return Status: 0
```

**Comment by realmazharhussain** [comment]
What is the output of the following commands?

```bash
flatpak run --command=sh io.github.realmazharhussain.GdmSettings
flatpak-spawn --host pkexec --version; echo Return Status: $?
exit
```

**Comment by RossComputerGuy** [comment]
```
Portal call failed: Failed to start command: 子プロセス“pkexec”の起動に失敗しました (No such file or directory)
Return Status: 1
```

**Comment by realmazharhussain** [comment]
It looks like I need to install NixOS myself to add support for it. I will do it someday. Keep the issue open.

In the meantime, try using the AppImage or installing the app from source and see if it works.

**Comment by michaelBelsanti** [comment]
It is failing because NixOS doesn't follow the FHS. From what I can tell, it searches in /usr/share, /usr/local/share, and HOST_DATA_DIRS. NixOS doesn't have /usr/local/share, because files are stored in the nix store at /nix/store. The application works if you set HOST_DATA_DIRS to `{gdm path}/share`. However, this can't be done reliably because the GDM path will change when it is updated. 

I am currently attempting to package this app in Nixpkgs where it can be wrapped with an environment variable that will set HOST_DATA_DIRS=`{gdm path}/share`.

**Comment by michaelBelsanti** [comment]
I'm actually not sure how, if at all, this app could work on Nix. For whatever reason, the file managing GDM's configuration is stored in the nix-store, which is immutable, meaning any changes you try to make with this app will fail. I'm trying to see if there are any workarounds that can be made.

**Comment by RossComputerGuy** [comment]
`/run/current-system/` kinda has what is running so that might help.

**Comment by michaelBelsanti** [comment]
I did check in there, and unfortunately there aren't any GDM or gnome-shell related files. Right now the only method I can think of would be using this app, exporting the config, and using an override to patch gdm with it.

**Comment by RossComputerGuy** [comment]
What about querying the nix store?

**Comment by michaelBelsanti** [comment]
Sorry I'm not sure exactly what you mean, query the store how?

**Comment by RossComputerGuy** [comment]
Using the `nix-store --query` command, or you could try using `systemctl` to find the service file which should point to the exec path of GDM.

**Comment by jtojnar** [comment]
Even if you find the resource file this way, you will not be able to replace it, since GDM looks for it in the immutable store. At least not without patching GDM to look for resources in some mutable location. And then you no longer have a guarantee that you system is in a consistent state when you update it.

Best way is getting the app export settings and then replacing the resource file during the build of your system something like this:

https://discourse.nixos.org/t/using-gdm-settings-on-nixos/23173/2?u=jtojnar

**Comment by michaelBelsanti** [comment]
Yes sorry, that's what I was trying to say. I was able to launch the application by locating GDM in the Nix store, but it isn't usable because of the immutability of the nix store. The only options are patching it when your system builds or replacing the file as jtojnar said.

This will only change if GDM adds a more sane configuration method, like reading from a gresource or other config file in /etc.

**Comment by Sudococommunity** [comment]
i have same error polkit is missing in my os too my os is parrot os 5.1 debian 10 based os here  is output of command given above pkexec --version; echo Return Status: $?
pkexec version 0.105
Return Status: 0
┌─[root@parrot]─[/flatpak run --command=sh io.github.realmazharhussain.GdmSettings
flatpak-spawn --host pkexec --version; echo Return Status: $?
exit
[📦 io.github.realmazharhussain.GdmSettings ~]# 
[📦 io.github.realmazharhussain.GdmSettings ~]# 
[📦 io.github.realmazharhussain.GdmSettings ~]# flatpak-spawn --host pkexec --version; echo Return Status: $?
Can't find bus: Error spawning command line “dbus-launch --autolaunch=3abeab6155ae42088a72c52638b449ff --binary-syntax --close-stderr”: Child process exited with code 1
Return Status: 1
[📦 io.github.realmazharhussain.GdmSettings ~]#

**Comment by realmazharhussain** [comment]
@Sudococommunity, the issue you are having is different. It seems like there is some problem with your flatpak or dbus setup.

If you have any more questions or comments about your issue, please, [open a new discussion topic](https://github.com/orgs/gdm-settings/discussions/new/choose) or a [new issue](https://github.com/gdm-settings/gdm-settings/issues/new/choose). Don't reply/comment in this one. This issue is for discussing gdm-settings on NixOS specifically.

Thank you.

**Comment by realmazharhussain** [comment]
Since this is not an actual GDM Settings issue but an external packaging/limitation issue, I am converting it into a discussion topic where everyone can discuss it freely without it cluttering the issue tracker.
